cpp/src/arrow/c/bridge.cc

```diff
@@ -2511,4 +2516,345 @@ Result<std::shared_ptr<ChunkedArray>> ImportDeviceChunkedArray(
   return ImportChunked</*IsDevice=*/true>(stream, mapper);
 }
 
+namespace {
+
+class AsyncRecordBatchIterator {
+ public:
+  struct State {
+    State(uint64_t queue_size, const DeviceMemoryMapper mapper)
+        : queue_size_{queue_size}, mapper_{std::move(mapper)} {}
+
+    const uint64_t queue_size_;
+    const DeviceMemoryMapper mapper_;
+    ArrowAsyncProducer* producer_;
+    DeviceAllocationType device_type_;
+
+    std::mutex mutex_;
+    std::shared_ptr<Schema> schema_;
+    std::condition_variable cv_;
+    std::queue<std::pair<ArrowAsyncTask, std::shared_ptr<KeyValueMetadata>>> batches_;
```

Comment:
Nit: instead of pair, could we use
```c++
    struct TaskWithMetadata {
      ArrowAsyncTask task;
      std::shared_ptr<KeyValueMetadata>> metadata;
    };
```

and rename this
```suggestion
    std::queue<TaskWithMetadata> tasks_;
```